Installation from source (v. 1.0.0) fails

**Using GCC**

```text
* installing *source* package ‘testthat’ ... (testthat_1.0.0.tar.gz)
** package ‘testthat’ successfully unpacked and MD5 sums checked
** libs
gcc -m64 -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c label.c -o label.o
gcc -m64 -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c reassign.c -o reassign.o
g++ -m64 -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c test-catch.cpp -o test-catch.o
In file included from ../inst/include/testthat.h:1:0,
                 from test-catch.cpp:1:
../inst/include/testthat/testthat.h:46:15: error: ‘std::testthat_exit_override’ has not been declared
 # define exit testthat_exit_override
               ^
RSite/R/etc/Makeconf:141: recipe for target 'test-catch.o' failed
make: *** [test-catch.o] Error 1
ERROR: compilation failed for package ‘testthat’
```


**Using Clang**

```text
* installing *source* package ‘testthat’ ...
** package ‘testthat’ successfully unpacked and MD5 sums checked
** libs
ccache clang -Qunused-arguments -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c label.c -o label.o
ccache clang -Qunused-arguments -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c reassign.c -o reassign.o
ccache clang++ -Qunused-arguments -I/usr/include/R -DNDEBUG -I../inst/include -DCOMPILING_TESTTHAT -I/usr/local/include    -fpic  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic  -c test-catch.cpp -o test-catch.o
In file included from test-catch.cpp:1:
In file included from ../inst/include/testthat.h:1:
In file included from ../inst/include/testthat/testthat.h:48:
In file included from ../inst/include/testthat/vendor/catch.h:481:
/bin/../lib/gcc/x86_64-redhat-linux/6.0.0/../../../../include/c++/6.0.0/stdlib.h:40:7: error: no member named 'testthat_exit_override' in namespace 'std'; did you mean '::testthat_exit_override'?
using std::testthat_exit_override;
      ^~~~~
../inst/include/testthat/testthat.h:40:24: note: '::testthat_exit_override' declared here
extern "C" inline void testthat_exit_override(int status) throw() {}
                       ^
1 error generated.
RSite/R/etc/Makeconf:141: recipe for target 'test-catch.o' failed
make: *** [test-catch.o] Error 1
ERROR: compilation failed for package ‘testthat’
```

```r
> devtools::session_info()
 setting  value                                      
 version  R version 3.2.4 Revised (2016-03-16 r70336)
 system   x86_64, linux-gnu                          
 ui       X11                                        
 language (EN)                                       
 collate  en_US.UTF-8

> R.version$platform
[1] "x86_64-redhat-linux-gnu"      
```


However, installation goes smoothly when 
```r
> devtools::session_info()          
 version  R version 3.2.3 (2015-12-10)
 system   x86_64, linux-gnu           
```

**OS (distribution)**

```bash
$ lsb_release -a
Distributor ID:	Fedora
```

**Edit: compilers' versions**
The GCC and Clang versions on the failed system differ from those on the successful one. 
```shell
$ gcc --version
Good vs Bad:
gcc (GCC) 5.3.1 20160406 (Red Hat 5.3.1-6) vs gcc (GCC) 6.0.0 20160406 (Red Hat 6.0.0-0.20)
$ clang --version
Good vs Bad:
clang version 3.7.0 (tags/RELEASE_370/final) vs clang version 3.8.0 (tags/RELEASE_380/final)
```

--- a/inst/include/testthat/testthat.h
+++ b/inst/include/testthat/testthat.h
@@ -34,22 +34,19 @@
 #  define CATCH_CONFIG_RUNNER
 # endif
 
-// Catch has calls to 'exit' on failure, which upset R CMD check.
-// We won't bump into them during normal test execution so just temporarily
-// hide it when we include 'catch'. Make sure we get 'exit' from the normal
-// places first, though.
-# include <cstddef> // std::size_t
-# include <cstdlib> // exit
 # include <cstdio>  // EOF
-extern "C" inline void testthat_exit_override(int status) throw() {}
 
 # ifdef __GNUC__
 #  pragma GCC diagnostic ignored "-Wparentheses"
 # endif
 
-# define exit testthat_exit_override
+// Catch has calls to 'exit' on failure, which upsets R CMD check.
+// We won't bump into them during normal test execution so just override
+// it in the Catch namespace before we include 'catch'.
+namespace Catch {
+inline void exit(int status) throw() {}
+};
 # include "vendor/catch.h"
-# undef exit
 
 // Implement an output stream that avoids writing to stdout / stderr.
 extern "C" void Rprintf(const char*, ...);

